---
# Grouping for Aggregation, Filtration, and Transformation
---

In [1]:
import numpy as np
import pandas as pd

## Defining an aggregation
we examine the flights dataset and perform the simplest aggregation involving only a single grouping column, a single aggregating column, and a single aggregating function.
We will find the average arrival delay for each airline.

In [2]:
flights = pd.read_csv('./flights.csv')
flights.head()

MONTH  DAY  WEEKDAY AIRLINE  ... SCHED_ARR ARR_DELAY  DIVERTED  CANCELLED
0      1    1        4      WN  ...      1905      65.0         0          0
1      1    1        4      UA  ...      1333     -13.0         0          0
2      1    1        4      MQ  ...      1453      35.0         0          0
3      1    1        4      AA  ...      1935      -7.0         0          0
4      1    1        4      WN  ...      2225      39.0         0          0

[5 rows x 14 columns]

Define the grouping columns (AIRLINE), aggregating columns (ARR_DELAY), and aggregating functions (mean). Place the grouping column in the .groupby method and then call the .agg method with a dictionary pairing the aggregating column with its aggregating function. If you pass in a dictionary, it returns back a DataFrame instance

In [3]:
(
    flights.groupby(['AIRLINE'])
    .agg({'ARR_DELAY': 'mean'})
)

ARR_DELAY
AIRLINE           
AA        5.542661
AS       -0.833333
B6        8.692593
DL        0.339691
EV        7.034580
F9       13.630651
HA        4.972973
MQ        6.860591
NK       18.436070
OO        7.593463
UA        7.765755
US        1.681105
VX        5.348884
WN        6.397353

Alternatively, we can place the aggregating column in the index operator and then pass the aggregating function as a string to `.agg`. This will return a Series:

In [8]:
(
    flights.groupby('AIRLINE')
    ['ARR_DELAY'].agg('mean')
 
)

AIRLINE
AA     5.542661
AS    -0.833333
B6     8.692593
DL     0.339691
EV     7.034580
F9    13.630651
HA     4.972973
MQ     6.860591
NK    18.436070
OO     7.593463
UA     7.765755
US     1.681105
VX     5.348884
WN     6.397353
Name: ARR_DELAY, dtype: float64

With numpy mean

In [5]:
(
    flights.groupby('AIRLINE')
    ['ARR_DELAY'].agg(np.mean)
    .head()
)

AIRLINE
AA    5.542661
AS   -0.833333
B6    8.692593
DL    0.339691
EV    7.034580
Name: ARR_DELAY, dtype: float64

It's possible to skip the agg method altogether in this case and use the code in text
method directly.

In [6]:
(
    flights.groupby('AIRLINE')
    ['ARR_DELAY'].mean()
    .head()
)

AIRLINE
AA    5.542661
AS   -0.833333
B6    8.692593
DL    0.339691
EV    7.034580
Name: ARR_DELAY, dtype: float64

In [7]:
flights.groupby('AIRLINE')['ARR_DELAY'].agg([np.std]).head()

std
AIRLINE           
AA       43.323160
AS       31.168354
B6       40.221718
DL       32.299471
EV       36.682336

## Grouping and aggregating with multiple columns and functions

As usual with any kind of grouping operation, it helps to identify the **three** components: 
-  the grouping columns, 
-  aggregating columns, 
-  and aggregating functions.

`.groupby` method by answering the following queries:
-  Finding the number of canceled flights for every airline per weekday
- Finding the number and percentage of canceled and diverted flights for every airline per weekday
-  For each origin and destination, finding the total number of flights, the number and percentage of canceled flights, and the average and variance of the airtime

In [15]:
flights.groupby(['AIRLINE', 'WEEKDAY'])['CANCELLED'].agg('sum')

AIRLINE  WEEKDAY
AA       1          41
         2           9
         3          16
         4          20
         5          18
                    ..
WN       3          18
         4          10
         5           7
         6          10
         7           7
Name: CANCELLED, Length: 98, dtype: int64

In [18]:
flights.groupby(['AIRLINE', 'WEEKDAY'])[['CANCELLED', 'DIVERTED']].agg(['sum', 'mean'])

CANCELLED           DIVERTED          
                      sum      mean      sum      mean
AIRLINE WEEKDAY                                       
AA      1              41  0.032106        6  0.004699
        2               9  0.007341        2  0.001631
        3              16  0.011949        2  0.001494
        4              20  0.015004        5  0.003751
        5              18  0.014151        1  0.000786
...                   ...       ...      ...       ...
WN      3              18  0.014118        2  0.001569
        4              10  0.007911        4  0.003165
        5               7  0.005828        0  0.000000
        6              10  0.010132        3  0.003040
        7               7  0.006066        3  0.002600

[98 rows x 4 columns]

In [19]:
(
    flights.groupby(['ORG_AIR', 'DEST_AIR'])
    .agg(
        {
            'CANCELLED':['sum', 'mean', 'size'],
            'AIR_TIME': ['mean', 'var']
        })
 )

CANCELLED                   AIR_TIME            
                       sum      mean size        mean         var
ORG_AIR DEST_AIR                                                 
ATL     ABE              0  0.000000   31   96.387097   45.778495
        ABQ              0  0.000000   16  170.500000   87.866667
        ABY              0  0.000000   19   28.578947    6.590643
        ACY              0  0.000000    6   91.333333   11.466667
        AEX              0  0.000000   40   78.725000   47.332692
...                    ...       ...  ...         ...         ...
SFO     SNA              4  0.032787  122   64.059322   11.338331
        STL              0  0.000000   20  198.900000  101.042105
        SUN              0  0.000000   10   78.000000   25.777778
        TUS              0  0.000000   20  100.200000   35.221053
        XNA              0  0.000000    2  173.500000    0.500000

[1130 rows x 5 columns]

To flatten the columns, we can use the `.to_flat_index` method

In [21]:
res = (
    flights.groupby(['ORG_AIR', 'DEST_AIR']).agg(
        {
            'CANCELLED':['sum', 'mean', 'size'],
            'AIR_TIME': ['mean', 'var']
        }
    )
)
# res.columns
res.columns = ['_'.join(x) for x in res.columns.to_flat_index()]
res

CANCELLED_sum  CANCELLED_mean  ...  AIR_TIME_mean  AIR_TIME_var
ORG_AIR DEST_AIR                                 ...                             
ATL     ABE                   0        0.000000  ...      96.387097     45.778495
        ABQ                   0        0.000000  ...     170.500000     87.866667
        ABY                   0        0.000000  ...      28.578947      6.590643
        ACY                   0        0.000000  ...      91.333333     11.466667
        AEX                   0        0.000000  ...      78.725000     47.332692
...                         ...             ...  ...            ...           ...
SFO     SNA                   4        0.032787  ...      64.059322     11.338331
        STL                   0        0.000000  ...     198.900000    101.042105
        SUN                   0        0.000000  ...      78.000000     25.777778
        TUS                   0        0.000000  ...     100.200000     35.221053
        XNA                   0        0.000000  ...     173.500000      0.500000

[1130 rows x 5 columns]

An eleguant method to flatten is to chaining with `.pipe` method

In [22]:
def flatten_cols(df):
  df.columns = ['_'.join(x) for x in df.columns.to_flat_index()]
  return df

In [25]:
res = (
    flights.groupby(['ORG_AIR', 'DEST_AIR']).agg(
        {
            'CANCELLED':['sum', 'mean', 'size'],
            'AIR_TIME': ['mean', 'var']
        }
    ).pipe(flatten_cols)
)

res

CANCELLED_sum  CANCELLED_mean  ...  AIR_TIME_mean  AIR_TIME_var
ORG_AIR DEST_AIR                                 ...                             
ATL     ABE                   0        0.000000  ...      96.387097     45.778495
        ABQ                   0        0.000000  ...     170.500000     87.866667
        ABY                   0        0.000000  ...      28.578947      6.590643
        ACY                   0        0.000000  ...      91.333333     11.466667
        AEX                   0        0.000000  ...      78.725000     47.332692
...                         ...             ...  ...            ...           ...
SFO     SNA                   4        0.032787  ...      64.059322     11.338331
        STL                   0        0.000000  ...     198.900000    101.042105
        SUN                   0        0.000000  ...      78.000000     25.777778
        TUS                   0        0.000000  ...     100.200000     35.221053
        XNA                   0        0.000000  ...     173.500000      0.500000

[1130 rows x 5 columns]

Be aware that when grouping with multiple columns, pandas creates a hierarchical index, or multi-index. In the preceding example, it returned 1,130 rows. However, if one of the columns that we group by is categorical (and has a category type, not an object type), then pandas will create a Cartesian product of all combinations for each level. In this case, it returns 2,710 rows. However, if you have categorical columns with higher cardinality, you can get many more values

In [26]:
res = (
    flights.assign(
        ORG_AIR = flights.ORG_AIR.astype('category')
    ).groupby(['ORG_AIR', 'DEST_AIR']).agg(
        {
            'CANCELLED':['sum', 'mean', 'size'],
            'AIR_TIME': ['mean', 'var']
        }
    ).pipe(flatten_cols)
)

In [27]:
res

CANCELLED_sum  CANCELLED_mean  ...  AIR_TIME_mean  AIR_TIME_var
ORG_AIR DEST_AIR                                 ...                             
ATL     ABE                   0             0.0  ...      96.387097     45.778495
        ABI                   0             NaN  ...            NaN           NaN
        ABQ                   0             0.0  ...     170.500000     87.866667
        ABR                   0             NaN  ...            NaN           NaN
        ABY                   0             0.0  ...      28.578947      6.590643
...                         ...             ...  ...            ...           ...
SFO     TYS                   0             NaN  ...            NaN           NaN
        VLD                   0             NaN  ...            NaN           NaN
        VPS                   0             NaN  ...            NaN           NaN
        XNA                   0             0.0  ...     173.500000      0.500000
        YUM                   0             NaN  ...            NaN           NaN

[2710 rows x 5 columns]

To remedy the combinatoric explosion, use the `observed=True` parameter. This makes the categorical group bys work like grouping with string types, and only shows the observed values and not the Cartesian product

In [28]:
res = (
    flights.assign(
        ORG_AIR = flights.ORG_AIR.astype('category')
    ).groupby(['ORG_AIR', 'DEST_AIR'], observed=True).agg(
        {
            'CANCELLED':['sum', 'mean', 'size'],
            'AIR_TIME': ['mean', 'var']
        }
    ).pipe(flatten_cols)
)

res

CANCELLED_sum  CANCELLED_mean  ...  AIR_TIME_mean  AIR_TIME_var
ORG_AIR DEST_AIR                                 ...                             
LAX     ABQ                   1        0.018182  ...      89.259259     29.403215
        ANC                   0        0.000000  ...     307.428571     78.952381
        ASE                   1        0.038462  ...     102.920000    102.243333
        ATL                   0        0.000000  ...     224.201149    127.155837
        AUS                   0        0.000000  ...     150.537500     57.897310
...                         ...             ...  ...            ...           ...
MSP     TTN                   1        0.125000  ...     124.428571     57.952381
        TUL                   0        0.000000  ...      91.611111     63.075163
        TUS                   0        0.000000  ...     176.000000     32.000000
        TVC                   0        0.000000  ...      56.600000     10.300000
        XNA                   0        0.000000  ...      90.642857    115.939560

[1130 rows x 5 columns]

## Removing the MultiIndex after grouping
MultiIndexes can happen in both the index and the columns. DataFrames with MultiIndexes are more difficult to navigate. The objective is to manipulate the index so that it has a single level and the column names are descriptive.

In [29]:
flights.head(2)

MONTH  DAY  WEEKDAY AIRLINE  ... SCHED_ARR ARR_DELAY  DIVERTED  CANCELLED
0      1    1        4      WN  ...      1905      65.0         0          0
1      1    1        4      UA  ...      1333     -13.0         0          0

[2 rows x 14 columns]

In [31]:
airline_info = (
    flights.groupby(['AIRLINE', 'WEEKDAY']).agg(
        {
            'DIST': ['sum', 'mean'],
            'ARR_DELAY': ['min', 'max']
        }
    )
    .astype(int)
)

airline_info

DIST       ARR_DELAY     
                     sum  mean       min  max
AIRLINE WEEKDAY                              
AA      1        1455386  1139       -60  551
        2        1358256  1107       -52  725
        3        1496665  1117       -45  473
        4        1452394  1089       -46  349
        5        1427749  1122       -41  732
...                  ...   ...       ...  ...
WN      3         997213   782       -38  262
        4        1024854   810       -52  284
        5         981036   816       -44  244
        6         823946   834       -41  290
        7         945679   819       -45  261

[98 rows x 4 columns]

In [32]:
print(airline_info.columns.get_level_values(0))
print()
print(airline_info.columns.get_level_values(1))

Index(['DIST', 'DIST', 'ARR_DELAY', 'ARR_DELAY'], dtype='object')

Index(['sum', 'mean', 'min', 'max'], dtype='object')


In [33]:
airline_info.columns = ['_'.join(x) for x in 
                        airline_info.columns.to_flat_index()]
airline_info.columns

Index(['DIST_sum', 'DIST_mean', 'ARR_DELAY_min', 'ARR_DELAY_max'], dtype='object')

A quick way to get rid of the row MultiIndex is to use the `.reset_index ` method:

In [34]:
airline_info.reset_index()

AIRLINE  WEEKDAY  DIST_sum  DIST_mean  ARR_DELAY_min  ARR_DELAY_max
0       AA        1   1455386       1139            -60            551
1       AA        2   1358256       1107            -52            725
2       AA        3   1496665       1117            -45            473
3       AA        4   1452394       1089            -46            349
4       AA        5   1427749       1122            -41            732
..     ...      ...       ...        ...            ...            ...
93      WN        3    997213        782            -38            262
94      WN        4   1024854        810            -52            284
95      WN        5    981036        816            -44            244
96      WN        6    823946        834            -41            290
97      WN        7    945679        819            -45            261

[98 rows x 6 columns]

In [35]:
airline_info = (
    flights.groupby(['AIRLINE', 'WEEKDAY']).agg(
        {
            'DIST':['sum', 'min'],
            'ARR_DELAY': ['min', 'max'],
        }
    )
    .astype(int)
    .pipe(flatten_cols)
    .reset_index()
)

airline_info

AIRLINE  WEEKDAY  DIST_sum  DIST_min  ARR_DELAY_min  ARR_DELAY_max
0       AA        1   1455386       175            -60            551
1       AA        2   1358256       175            -52            725
2       AA        3   1496665       175            -45            473
3       AA        4   1452394       175            -46            349
4       AA        5   1427749       175            -41            732
..     ...      ...       ...       ...            ...            ...
93      WN        3    997213       197            -38            262
94      WN        4   1024854       197            -52            284
95      WN        5    981036       197            -44            244
96      WN        6    823946       197            -41            290
97      WN        7    945679       197            -45            261

[98 rows x 6 columns]

By default, at the end of a groupby operation, pandas puts all of the grouping columns in the index. The `as_index` parameter in the `.groupby` method can be set to False to avoid this behavior. You can chain the `.reset_index` method after grouping to get the same effect 

In [36]:
(
    flights.groupby('AIRLINE', as_index=False)['DIST']
    .agg('mean')
    .round(0)
)

AIRLINE    DIST
0       AA  1114.0
1       AS  1066.0
2       B6  1772.0
3       DL   866.0
4       EV   460.0
5       F9   970.0
6       HA  2615.0
7       MQ   404.0
8       NK  1047.0
9       OO   511.0
10      UA  1231.0
11      US  1181.0
12      VX  1240.0
13      WN   810.0

In [38]:
airline_info = (
    flights.groupby(['AIRLINE', 'WEEKDAY'], as_index=False).agg(
        {
            'DIST':['sum', 'min'],
            'ARR_DELAY': ['min', 'max'],
        }
    )
    # .astype(int)
    .pipe(flatten_cols)
    # .reset_index()
)

airline_info

AIRLINE_  WEEKDAY_  DIST_sum  DIST_min  ARR_DELAY_min  ARR_DELAY_max
0        AA         1   1455386       175          -60.0          551.0
1        AA         2   1358256       175          -52.0          725.0
2        AA         3   1496665       175          -45.0          473.0
3        AA         4   1452394       175          -46.0          349.0
4        AA         5   1427749       175          -41.0          732.0
..      ...       ...       ...       ...            ...            ...
93       WN         3    997213       197          -38.0          262.0
94       WN         4   1024854       197          -52.0          284.0
95       WN         5    981036       197          -44.0          244.0
96       WN         6    823946       197          -41.0          290.0
97       WN         7    945679       197          -45.0          261.0

[98 rows x 6 columns]